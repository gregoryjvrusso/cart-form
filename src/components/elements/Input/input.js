import React from "react";
import { func, string, object } from "prop-types";
import Box from "../Box";
import { Container, Label, TextMessage } from "./input.styles";

const Input = ({
  name,
  textLabel,
  mask,
  onInputBlur,
  onInputChange,
  messageError,
  value,
  setValue,
  setError,
  error,
  ...props
}) => (
  <Box width={"100%"} pt={3}>
    <Container
      type="input"
      placeholder={textLabel}
      data-testid="input"
      name={name}
      id={name}
      required
      value={value}
      mask={mask}
      setValue={setValue}
      onBlur={(e) => onInputBlur(e, error, setError, name, props.regex)}
      onChange={(e) => onInputChange(e, setValue)}
      {...props}
      maskChar={null}
    />
    <Label htmlFor={name} {...props}>
      {textLabel}
    </Label>
    {error[name] && <TextMessage>{messageError}</TextMessage>}
  </Box>
);

Input.propTypes = {
  name: string.isRequired,
  textLabel: string.isRequired,
  mask: string,
  messageError: string,
  onInputBlur: func.isRequired,
  onInputChange: func.isRequired,
  value: string,
  setValue: func,
  setError: func,
  error: object,
};

Input.defaultProps = {
  mask: undefined,
  messageError: string,
  value: "",
  setValue: undefined,
  error: {},
  setError: undefined,
};

Input.displayName = "Input";

export default Input;
