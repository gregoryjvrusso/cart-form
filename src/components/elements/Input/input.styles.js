import styled from "styled-components";
import { theme } from "styled-tools";
import InputMask from "react-input-mask";
import Text from "../Text";

export const Container = styled(InputMask)`
  background-color: transparent;
  border: ${theme("spaces.0")};
  border-bottom: 2px solid
    ${(props) =>
      props.error ? theme(`colors.texts.red`) : theme(`colors.texts.gray`)};
  font-size: ${theme("fontSizes.2")}px;
  width: 100%;

  &::placeholder {
    color: transparent;
  }
`;

export const Label = styled.label`
  font-size: ${theme("fontSizes.2")}px;
  color: ${theme(`colors.texts.gray`)};
  left: ${theme("spaces.2")}px;
  pointer-events: none;
  position: absolute;
  top: ${theme("spaces.4")}px;
  transition: 0.3s ease all;
  

  ${Container}:focus ~ &,
  ${Container}:valid ~ & {
    font-size: ${theme("fontSizes.0")}px;
    opacity: 0.6;
    left: ${theme("spaces.2")}px;
    top: ${theme("spaces.1")}px;
  }
`;

export const TextMessage = styled(Text)`
  font-family: ${theme("fontFamily.0")};
  font-size: ${theme("fontSizes.0")}px;
  color: ${theme(`colors.texts.red`)};
  margin-top: ${theme("spaces.0")};
`;
