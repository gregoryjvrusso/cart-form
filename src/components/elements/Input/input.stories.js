import React, { useState } from "react";
import { storiesOf } from "@storybook/react";
import { ThemeProvider } from "styled-components";
import theme from "../../../common/theme";

import Input from "../Input";

storiesOf("Elements/Input", module).add("default", () => {
  const [value, setValue] = useState("");
  const [error, setError] = useState(false);
  const regex = RegExp(/\b[A-Za-zÀ-ú][A-Za-zÀ-ú]+,?\s[A-Za-zÀ-ú][A-Za-zÀ-ú]{2,19}\b/gi);

  const mockOnInputChange = (e) => {
    setValue(e.target.value);
    regex.test(e.target.value) ? setError(false) : setError(true);
  };
  return (
    <ThemeProvider theme={theme}>
      <Input
        name="nome"
        textLabel="Nome"
        value={value}
        error={error}
        messageError="Insira seu nome completo"
        onInputChange={mockOnInputChange}
      />
    </ThemeProvider>
  );
});
