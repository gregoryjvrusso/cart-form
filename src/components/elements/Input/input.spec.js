import React from "react";
import { fireEvent, render } from "@testing-library/react";
import Input from "../Input";

describe("Input component", () => {
  it("deve conseguir renderizar [snapshot]", async () => {
    const onChangeMock = jest.fn();
    const onBlurMock = jest.fn();
    const { container } = render(
      <Input
        name="nome"
        textLabel="Nome"
        messageError="Insira seu nome completo"
        onInputChange={onChangeMock}
        onInputBlur={onBlurMock}
      />
    );
    expect(container.firstChild).toMatchSnapshot();
  });
  it("deve conseguir renderizar [snapshot]", async () => {
    const onChangeMock = jest.fn();
    const onBlurMock = jest.fn();
    const { getByTestId } = render(
      <Input
        name="nome"
        textLabel="Nome"
        messageError="Insira seu nome completo"
        onInputChange={onChangeMock}
        onInputBlur={onBlurMock}
      />
    );
    const input = getByTestId("input");
    expect(input.value).toBe('');
    fireEvent.change(input, { target: { value: 'Nome' } });
    expect(input.value).toBe('Nome');
  });
});
