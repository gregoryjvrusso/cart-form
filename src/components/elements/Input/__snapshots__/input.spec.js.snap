// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input component deve conseguir renderizar [snapshot] 1`] = `
<div
  class="sc-AxjAm jpulAN"
  data-testid="box"
  width="100%"
>
  <input
    class="sc-AxiKw eQYNKT"
    data-testid="input"
    id="nome"
    name="nome"
    placeholder="Nome"
    required=""
    type="input"
    value=""
  />
  <label
    class="sc-AxhCb bssWpm"
    for="nome"
  >
    Nome
  </label>
</div>
`;
